### DISCUSSION TITLE
Overrule timer with switch input

### DISCUSSION THREAD

**Discussion by jan-gerard** [arendst/Tasmota#13211] [Q&A] [answered]
Can I make a timer that can be overruled by a switch input? I have a digital input as switch. When 
I am connecting a Sonoff Basic to my floor heating pump, controlled by a digitial input wired via an optocoupler to the floor-heating circuit valve. When the valve opens, the pump needs to start, and when closed, it should stop after a delay. Furthermore, at 4AM the pump needs to be switched on for 5 minutes by a timer function. 
So far, so good. But I don't want the timer to switch off the pump when the valve happens to be open. So when the digital input is low, the timer may switch on and off the pump, but when high, the timer should not be able to switch it off.
I see that timers cannot be used with rules, so how can I prevent the timer to switch off the output?

**Comment by sfromis** [comment]
Timers **can** actually be used with rules, when you configure a timer for that, you have a rule trigger when the timer fires. This allows you to do the logic you want in a rule.

**Comment by jan-gerard** [comment]
I am trying to, but I'm already stuck at using the timer to switch (and I even didn't include conditional statements yet).
`Timer1 {"Enable":1,"Mode":0,"Time":"17:26","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}`
`Rule1 on clock#Timer=1 do power1 toggle endon`
Could there be some SetOption commands that is set incorrectly? 

The final version should be something like
```
Backlog switchmode1 2; rule1 1
Rule1 ON switch1#state=1 DO Backlog Power1 on; ruletimer1 0 ENDON
ON switch1#state=0 DO ruletimer1 300 ENDON
ON rules#timer=1 DO Power1 0 ENDON
Timer1 {"Enable":1,"Time":"04:500,"Window":0,"Days":"SMTWTFS","Repeat":1,"Output":1,"Action":3}
Rule2 ON Clock#Timer DO Backlog IF (switch1#state==0) Power1 1; ruletimer2 300 ENDON
ON rules#timer2 DO IF (switch1#state==0) Power1 0 ENDIF ENDON
```
But I don't get to that point, as simply switching by timer isn't working.

**Reply by ascillato** [reply]
Please check:

* That the timer is enabled and the output is set to rule
![image](https://user-images.githubusercontent.com/35405447/134777773-1f30a0bf-83fb-4659-843e-c995aa7dff2e.png)

* That the rule is ON (`rule` command to check its status)

* That your `timezone` is your real timezone

**Reply by sfromis** [reply]
Also note that the top tick mark (or command `Timers`) for all timers is sometimes overlooked

**Reply by barbudor** [reply]
`Rule2 ON Clock#Timer DO ...`
If this is a proper copy/paste, you are missing `=1`

**Reply by jan-gerard** [reply]
Thanks for your replies. One more thing I forgot was to activate my rule 2 with `rule2 1`. Most of it is working, except for the IF statement. If I remove that, everything is fine. 
Now I realize that the pre-compiled Tasmota binaries don't support conditional statements. I read that BREAK could be used to mimic that, but I don't succeed in what I want to achieve:
**Have a timer to switch my output on and off at a given time, unless the switch input is high; then the output should remain on.**
Any idea how I could achieve that without the IF statement? I'd like to avoid having to compile everything myself.

```
TimeZone 99
TimeDST 0,0,3,1,2,120
TimeSTD 0,0,10,1,3,60
Timer1 {"Enable":1,"Mode":0,"Time":"23:42","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}
Backlog switchmode1 2; rule1 1
Rule1 ON switch1#state=1 DO Backlog Power1 on; ruletimer1 0 ENDON
ON switch1#state=0 DO ruletimer1 300 ENDON
ON rules#timer=1 DO Power1 off ENDON
Rule1 1
Rule2 ON clock#Timer=1 DO IF (switch1#state=0) Backlog Power1 on; ruletimer2 300 ENDIF ENDON
ON rules#timer=2 DO IF (switch1#state=0) Power1 off ENDIF ENDON
Rule2 1
```

**Comment by jan-gerard** [comment, marked as answer]
[SOLVED] I found the solution: use variables. I declare Var1 as 0 (off) at start, then if the pump is switched on I set it to 1, and when switched off, I set it to 0. Then, in the rule that processes the clock timer, I set the pump output first to 1, and after the time has finished, I set it to Var1, such that if it wasn't switched on by the heater, it will be off, but when the heater tells it to be on (switch input), it will remain on. No conditional statements needed this way.

`Backlog switchmode1 2; rule1 1; rule2 1; Var1 0; TimeZone 99; TimeDST 0,0,3,1,2,120; TimeSTD 0,0,10,1,3,60`
`Timer1 {"Enable":1,"Mode":0,"Time":"04:00","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}`
```
Rule1 ON switch1#state=1 DO Backlog Power1 on; ruletimer1 0; Var1 1 ENDON
ON switch1#state=0 DO ruletimer1 300 ENDON
ON rules#timer=1 DO Backlog Power1 off; Var1 0 ENDON
```
```
Rule2 ON clock#Timer=1 DO Backlog Power1 1; ruletimer2 300 ENDIF ENDON
ON rules#timer=2 DO Power1 %Var1% ENDIF ENDON
```
